ext/storage_sources/s3_store/s3_storage_source.cpp

```diff
@@ -175,120 +146,62 @@ S3SourcePath(const std::string &dir, const std::string &name)
+    else {
+        if (fileExists)
+            s3->log->LogDebugMessage("S3FileExistsAndGetSize: Found file in S3.");
+        else
+            s3->log->LogDebugMessage("S3FileExistsAndGetSize: File not found in S3.");
+    }
 
-// Checks whether the given file exists in the cache.
-static bool
-S3CacheExists(WT_FILE_SYSTEM *fileSystem, const std::string &name)
-{
-    const std::string path = S3CachePath(((S3FileSystem *)fileSystem)->cacheDir, name);
-    return (LocalFileExists(path));
-}
+    if (*fileExists && size != NULL)
+        *size = objectSize;
 
-// Checks whether a file corresponding to the provided path exists locally.
-static bool
-LocalFileExists(const std::string &path)
-{
-    std::ifstream f(path);
-    return (f.good());
+    return (ret);
 }
 
-// Return a copy of a directory name after verifying that it is a directory.
 static int
-S3GetDirectory(const S3Storage &s3, const std::string &home, const std::string &name, bool create,
-  std::string &copy)
+S3FileExists(WT_FILE_SYSTEM *fileSystem, WT_SESSION *session, const char *name, bool *fileExists)
 {
-    // copy must be initialised before the function returns.
-    copy = "";
-
-    int ret;
-    std::string dirName;
-
-    // For relative pathnames, the path is considered to be relative to the home directory.
-    if (name[0] == '/')
-        dirName = name;
-    else
-        dirName = home + "/" + name;
-
-    // Use filesystem status to find if directory exists.
-    std::error_code ec;
-    std::filesystem::file_status status = std::filesystem::status(dirName.c_str(), ec);
-
-    if (!std::filesystem::exists(status) && create) {
-        try {
-            std::filesystem::create_directory(dirName.c_str());
-            std::filesystem::permissions(dirName.c_str(), std::filesystem::perms::all);
-        } catch (std::filesystem::filesystem_error const &e) {
-            s3.log->LogErrorMessage(std::string("S3GetDirectory: ") + e.what());
-        }
-
-        s3.log->LogDebugMessage("S3GetDirectory: Successfully created directory.");
-    }
-
-    status = std::filesystem::status(dirName.c_str(), ec);
-    ret = ec.value();
-
-    if (ret != 0) {
-        s3.log->LogErrorMessage("S3GetDirectory: No such file or directory");
-    } else if (!std::filesystem::is_directory(status)) {
-        s3.log->LogErrorMessage("S3GetDirectory: invalid directory name");
-        ret = EINVAL;
-    }
-
-    copy = dirName;
-    return (ret);
+    return (S3FileExistsAndGetSize(fileSystem, session, name, fileExists, NULL));
 }
 
 // File handle close.
 static int
 S3FileClose(WT_FILE_HANDLE *fileHandle, WT_SESSION *session)
 {
-    int ret = 0;
     S3FileHandle *s3FileHandle = (S3FileHandle *)fileHandle;
     S3Storage *s3 = s3FileHandle->storage;
-    WT_FILE_HANDLE *wtFileHandle = s3FileHandle->wtFileHandle;
+
+    // If there are other active instances of the file being open, do not close file handle.
+    if (--s3FileHandle->referenceCount != 0)
```

Comment:
Are file handles shared between threads? This reference count might need to be an atomic.

Reply:
I have changed it to an atomic [12ff933](https://github.com/wiredtiger/wiredtiger/pull/9708/commits/12ff9339c1e48e451f1f053627295e0c4ddb4836)